### PR TITLE
Serialiser_Engine: Fix issue serialising Immutable objects with properties overridden from base class.

### LINKED ISSUE #1796: Serialiser_Engine: Serialisation of objects fail if a Property has been overridden
#### Description:

If a virtual property on a base obejct, for example `Name` on the BHoMObject has been overridden, the object fail to serialise and deserialise. Might only be a problem with IImuttable objects, but needs to be investigated

#### Steps to reproduce:

Try overriding for example Name on an object and run to and from Json.

#### Test file(s):

Can be tested with

https://burohappold.sharepoint.com/:f:/s/BHoM/Esd8tkJZRzdAu30SbjtwfKwBppMw3IL5wEdOX1ddhQSCyg?e=cFuicU

while on the following branch: https://github.com/BHoM/BHoM/pull/879

### PR DESCRIPTION
**Pull request by IsakNaslundBh**
### Issues addressed by this PR

Closes #1796 

Check if property is overridden before registering it to class map for immutable objects

This is solving issues found while testing https://github.com/BHoM/BHoM/pull/879 where the sections failed to deserialise after overriding the `Name` property. 

### Test files

can be tested together with https://github.com/BHoM/BHoM/pull/879 

https://burohappold.sharepoint.com/:f:/s/BHoM/Esd8tkJZRzdAu30SbjtwfKwBppMw3IL5wEdOX1ddhQSCyg?e=eEB1rV

and general test all file in here:

https://burohappold.sharepoint.com/:f:/s/BHoM/EobV-zo3k9NCrXrp9WsnwF4BIfg0V4Xx520ZcZL94n2eEQ?e=06umbh

### Additional comments

Issue seem to be resolved, and all tests I have done work, but would want @adecler to have a look at this before merged. 

Also, limitation found with the fix:

If a property is overridden, it can NOT be included in the constructor for the immutable object. If it is, the serialisation will fail!

### CODE REVIEW

**Comment by IsakNaslundBh** [comment]
> Tested and works fine.
> 
> I would like to see a null check on `getMethod` though. I know the BHoM doesn't allow properties that are setters only, but this would throw a very unwelcome exception if someone ever create such object.

good catch. will fix!

**Comment by IsakNaslundBh** [comment]
Done, had to update from master to get it to build as BHoM had been merged.

File: Serialiser_Engine/Objects/MemberMapConventions/ImmutableBHoMClassMapConvention.cs [hunk 1]
```diff
@@ -103,6 +103,25 @@ public void Apply(BsonClassMap classMap)
...
+            var getMethod = property.GetGetMethod(false);
+            if (getMethod != null)
+            {
+                return getMethod.GetBaseDefinition() != getMethod;
+            }
+            else
+            {
+                //This case should not happen for compliant BHoMObjects, as all properties should have getters, but adding to be safe.
+                var setMethod = property.GetSetMethod(false);
+                return setMethod.GetBaseDefinition() != getMethod;
```
Comment (adecler) [inline Serialiser_Engine/Objects/MemberMapConventions/ImmutableBHoMClassMapConvention.cs:120]: Ouch! should be `!= setMethod`

Comment (IsakNaslundBh) [reply Serialiser_Engine/Objects/MemberMapConventions/ImmutableBHoMClassMapConvention.cs:120]: 🤦

**Pull request merged**